client.go

```diff
@@ -9,28 +10,52 @@ import (
 // the protocol specified for the handshake.
 var ErrNotSupported = errors.New("protocol not supported")
 
+// ErrNoProtocols is the error returned when the no protocols have been
+// specified.
+var ErrNoProtocols = errors.New("no protocols specified")
+
 // SelectProtoOrFail performs the initial multistream handshake
 // to inform the muxer of the protocol that will be used to communicate
 // on this ReadWriteCloser. It returns an error if, for example,
 // the muxer does not know how to handle this protocol.
 func SelectProtoOrFail(proto string, rwc io.ReadWriteCloser) error {
-	err := handshake(rwc)
-	if err != nil {
-		return err
+	errCh := make(chan error, 1)
+	go func() {
+		var buf bytes.Buffer
+		delimWrite(&buf, []byte(ProtocolID))
+		delimWrite(&buf, []byte(proto))
+		_, err := io.Copy(rwc, &buf)
+		errCh <- err
+	}()
+	// We have to read *both* errors.
+	err1 := readMultistreamHeader(rwc)
+	err2 := readProto(proto, rwc)
+	if werr := <-errCh; werr != nil {
+		return werr
 	}
-
-	return trySelect(proto, rwc)
+	if err1 != nil {
+		return err1
+	}
+	if err2 != nil {
+		return err2
+	}
+	return nil
 }
 
 // SelectOneOf will perform handshakes with the protocols on the given slice
 // until it finds one which is supported by the muxer.
 func SelectOneOf(protos []string, rwc io.ReadWriteCloser) (string, error) {
-	err := handshake(rwc)
-	if err != nil {
+	if len(protos) == 0 {
+		return "", ErrNoProtocols
+	}
+	switch err := SelectProtoOrFail(protos[0], rwc); err {
```

Comment:
why isn't this part of the loop below?

Reply:
I'm combining the `/multistream/1.0` handshake when negotiating the first protocol. I'll leave a comment.

Reply:
Done. This should drop us from two round trips to 1.

Reply:
ok, good.